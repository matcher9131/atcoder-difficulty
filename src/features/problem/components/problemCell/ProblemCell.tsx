import clsx from "clsx";
import type { ReactNode } from "react";

export type ProblemCellProps = {
    readonly iconHref: string;
    readonly fillColor: string;
    readonly difficulty: string;
    readonly displayName: string;
    readonly textColor: string;
    readonly linkHref: string;
};

export const ProblemCell = ({
    fillColor,
    iconHref,
    difficulty,
    displayName,
    textColor,
    linkHref,
}: ProblemCellProps): ReactNode => {
    return (
        <td className="px-3 py-6 gap-x-1">
            <div className="flex items-baseline">
                <span className="relative">
                    <svg role="img" aria-label="難易度アイコン" className={clsx("w-6", "h-4", fillColor)}>
                        <use href={iconHref} />
                    </svg>
                    <div className="absolute top-4 w-6 text-center text-[.5em]">{difficulty}</div>
                </span>
                <a href={linkHref} target="_blank" rel="noreferrer" className={clsx("truncate", textColor)}>
                    {displayName}
                </a>
            </div>
        </td>
    );
};
