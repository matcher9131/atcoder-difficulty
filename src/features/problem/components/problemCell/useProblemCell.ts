import { useAtom } from "jotai";
import { problemSelector } from "../../dict/problems";
import { getFillColor, getTextColor } from "./functions";
import type { ProblemCellProps } from "./ProblemCell";

export const useProblemCell = (problemId: string): ProblemCellProps => {
    const [problem] = useAtom(problemSelector(problemId));
    const [, difficulty] = problem.d ?? [null];
    const level = difficulty != null ? (difficulty >= 3200 ? 4 : (Math.floor(difficulty / 100) % 4) + 1) : 0;
    const iconHref = `/resources/up_arrow_${level}.svg#up_arrow_${level}`;
    const fillColor = difficulty != null ? getFillColor(difficulty) : "";
    const textColor = difficulty != null ? getTextColor(difficulty) : "";
    const contestName = problemId.substring(0, problemId.lastIndexOf("_"));
    const linkHref = `https://atcoder.jp/contests/${contestName}/tasks/${problemId}`;
    return {
        fillColor,
        iconHref,
        difficulty: difficulty != null ? `${difficulty}` : "NaN",
        displayName: problem.n,
        textColor,
        linkHref,
    };
};
