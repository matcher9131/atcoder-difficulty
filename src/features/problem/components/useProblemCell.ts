import { useAtom } from "jotai";
import { problemSelector } from "../dict/problems";
import { getFillColor, getTextColor } from "./functions";
import type { ProblemCellProps } from "./ProblemCell";

export const useProblemCell = (problemId: string): ProblemCellProps => {
    const [problem] = useAtom(problemSelector(problemId));
    const { difficulty, displayName } = problem;
    const level =
        difficulty >= 3200 ? 4 : (Math.floor(difficulty / 50) % 4) + 1;
    const iconHref = `/resources/up_arrow_${level}.svg#up_arrow_${level}`;
    const fillColor = getFillColor(difficulty);
    const textColor = getTextColor(difficulty);
    const contestName = problemId.substring(0, problemId.lastIndexOf("_"));
    const linkHref = `https://atcoder.jp/contests/${contestName}/tasks/${problemId}`;
    return {
        fillColor,
        iconHref,
        difficulty,
        displayName,
        textColor,
        linkHref,
    };
};
