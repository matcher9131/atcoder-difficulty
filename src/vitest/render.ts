import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
