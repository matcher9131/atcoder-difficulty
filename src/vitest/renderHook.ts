import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
