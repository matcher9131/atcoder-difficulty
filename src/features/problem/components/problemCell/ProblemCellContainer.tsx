import { ProblemCell } from "./ProblemCell";
import { useProblemCell } from "./useProblemCell";

type ProblemCellContainerProps = {
    readonly problemId: string;
    readonly showsParameters: boolean;
    readonly showsProblemIndex: boolean;
};

export const ProblemCellContainer = ({ problemId, showsParameters, showsProblemIndex }: ProblemCellContainerProps) => {
    const props = useProblemCell(problemId, showsParameters, showsProblemIndex);
    return <ProblemCell {...props} />;
};
