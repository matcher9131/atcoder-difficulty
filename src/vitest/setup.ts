import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
