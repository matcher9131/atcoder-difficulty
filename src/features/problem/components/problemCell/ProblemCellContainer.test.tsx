import { describe, it, expect, vi, beforeEach } from "vitest";
import { render } from "../../../../vitest/render";
import { ProblemCellContainer } from "./ProblemCellContainer";
import { useProblemCell } from "./useProblemCell";
import { ProblemCell } from "./ProblemCell";

// src/features/problem/components/problemCell/ProblemCellContainer.test.tsx

// Mock ProblemCell and useProblemCell
const mockProps = {
    title: "Mock Problem",
    difficulty: 1234,
    url: "https://mockproblem.com",
    index: "A",
    parameters: { timeLimit: 2, memoryLimit: 256 },
    onOpenGraphClick: vi.fn(),
};

vi.mock("./useProblemCell", () => ({
    useProblemCell: vi.fn(),
}));

vi.mock("./ProblemCell", () => ({
    ProblemCell: vi.fn(() => <div data-testid="problem-cell" />),
}));

describe("ProblemCellContainer", () => {
    const problemId = "p123";
    const showsParameters = true;
    const showsProblemIndex = false;
    const showsOpenGraphButton = true;

    beforeEach(() => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useProblemCell as any).mockClear();
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (ProblemCell as any).mockClear();
    });

    it("calls useProblemCell with correct arguments", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useProblemCell as any).mockReturnValue(mockProps);
        render(
            <ProblemCellContainer
                problemId={problemId}
                showsParameters={showsParameters}
                showsProblemIndex={showsProblemIndex}
                showsOpenGraphButton={showsOpenGraphButton}
            />,
        );
        expect(useProblemCell).toHaveBeenCalledWith(
            problemId,
            showsParameters,
            showsProblemIndex,
            showsOpenGraphButton,
        );
    });

    it("renders ProblemCell with props from hook", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useProblemCell as any).mockReturnValue(mockProps);
        render(
            <ProblemCellContainer
                problemId={problemId}
                showsParameters={showsParameters}
                showsProblemIndex={showsProblemIndex}
                showsOpenGraphButton={showsOpenGraphButton}
            />,
        );
        expect(ProblemCell).toHaveBeenCalledWith(mockProps, undefined);
    });

    it("matches snapshot", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useProblemCell as any).mockReturnValue(mockProps);
        const { container } = render(
            <ProblemCellContainer
                problemId={problemId}
                showsParameters={showsParameters}
                showsProblemIndex={showsProblemIndex}
                showsOpenGraphButton={showsOpenGraphButton}
            />,
        );
        expect(container).toMatchSnapshot();
    });
});
