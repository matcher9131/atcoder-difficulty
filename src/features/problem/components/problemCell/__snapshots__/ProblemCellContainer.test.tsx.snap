// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProblemCellContainer > matches snapshot 1`] = `
<div>
  <div
    data-testid="problem-cell"
  />
</div>
`;
